doc/subsystems/shell/shell.rst

```diff
@@ -0,0 +1,294 @@
+available subcommands for the currently entered command level.
+For example, when the cursor is positioned at the beginning of the command
+line and the Tab key is pressed, the user will see all root (level 0)
+commands:
+
+.. code-block:: none
+
+	  clear  demo  shell  history  log  resize  version
+
+
+.. note::
+	To view the subcommands that are available for a specific command, you must first type a space after this command and then hit Tab.
+
+These commands are registered by various modules, for example:
+
+* ``clear``, ``shell``, ``history``, and ``resize`` are built-in commands which have been registered by :file:`subsys/shell/.shell.c` ,
+* ``demo`` and ``version`` have been registered in example code above by main.c,
+* log has been registered by :file:`subsys/logging/log_cmds.c`
+
+Then, if a user types a ``demo`` command and presses the Tab key, the shell
+will only print the subcommands registered for this command:
+
+.. code-block:: none
+
+	  params  ping
+
+
+Commands
+********
+
+shell commands are organized in a tree structure and grouped into the following
+types:
+
+* Root command (level 0): Kept in the section variable area.
+* Static subcommand (level > 0): Number and syntax must be known during compile time. Kept in the software module.
+* Dynamic subcommand (level > 0): Number and syntax does not need to be known during compile time. Created in the software module.
+ 
+Creating commands
+=================
+
+Use the following macros for adding shell commands:
+
+* :c:macro:`SHELL_CMD_REGISTER` - Create root command. Each and every root command shall have unique syntax.
+* :c:macro:`SHELL_CMD` - Initialize a command.
+* :c:macro:`SHELL_CREATE_STATIC_SUBCMD_SET` - Create a static subcommands array. Static subcommands must be added in alphabetical order to ensure correct smart completion.
+* :c:macro:`SHELL_SUBCMD_SET_END` - shall be placed as last in :c:macro:`SHELL_CREATE_STATIC_SUBCMD_SET` macro.
+* :c:macro:`SHELL_CREATE_DYNAMIC_CMD` - Create a dynamic subcommands array. Dynamic subcommands must be returned in alphabetical order to ensure correct smart completion.
+
+Commands execution
+==================
+
+Each command or subcommand on each level can either have or not have a handler.
+The shell executes the handler that is found deepest in the command tree and
+further subcommands (without a handler) are passed as arguments.
+Characters within parentheses are treated as one argument. Each command or
+subcommand without a handler can be treated as an argument, or it can still
+have a subcommand with a handler and there is no problem for the user
+to execute it.
+
+Built-in commands
+=================
+
+* ``clear`` - Clears the screen.
+* ``shell`` - Useful shell-related commands like:
+	* ``echo`` - Toggles shell echo.
+        * ``colors`` - Toggles colored syntax. This might be helpful in case of Bluetooth shell to limit the amount of transferred bytes.
```

Comment:
keep line length <80 chars.  For bullet lists, subsequent lines should be indented to the first non-blank character:

    * ``colors`` - Toggles colored syntax.  This might be helpful for a Bluetooth
      transport shell to avoid transferring unnecessary bytes.